rust/agama-lib/src/utils/file_format.rs

```diff
@@ -0,0 +1,135 @@
+// Software Foundation; either version 2 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+//! File format detection for Agama.
+//!
+//! It implements a simple API to detect the file formats that are relevent for Agama.
+
+use std::{
+    io::Write,
+    path::Path,
+    process::{Command, Stdio},
+};
+
+/// Relevant file formats for Agama.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum FileFormat {
+    Json,
+    Jsonnet,
+    Script,
+    Unknown,
+}
+
+const JSONNETFMT_BIN: &str = "jsonnetfmt";
+
+impl FileFormat {
+    /// Tries to guess the file format from the content of a file.
+    pub fn from_file<P: AsRef<Path>>(file_path: P) -> Result<Self, std::io::Error> {
+        let content = std::fs::read_to_string(file_path)?;
+        Ok(Self::from_string(&content))
+    }
+
+    /// Tries to guess the file format from a string.
+    pub fn from_string(content: &str) -> Self {
+        if Self::is_json(content) {
+            return Self::Json;
+        } else if Self::is_jsonnet(content) {
+            return Self::Jsonnet;
+        } else if Self::is_script(content) {
+            return Self::Script;
+        }
+
+        Self::Unknown
+    }
+
+    /// Whether the format is JSON.
+    ///
+    /// It tries to parse the content as JSON and returns `true` if it succeeds.
+    fn is_json(content: &str) -> bool {
+        let json = serde_json::from_str::<serde_json::Value>(content);
+        json.is_ok()
+    }
+
+    /// Whether the format is Jsonnet.
+    ///
+    /// It tries to process the content with the jsonnetfmt tool and returns `true` if it succeeds.
+    fn is_jsonnet(content: &str) -> bool {
+        let child = Command::new(JSONNETFMT_BIN)
+            .args(["-"])
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn();
+
+        let Ok(mut child) = child else {
+            return false;
+        };
+
+        let Some(mut stdin) = child.stdin.take() else {
+            return false;
+        };
+
+        stdin
+            .write_all(content.as_bytes())
+            .expect("Failed to write to stdin");
+        drop(stdin);
+
+        child.wait().is_ok_and(|s| s.success())
+    }
+
+    /// Whether is is a script.
+    ///
+    /// It returns `true` if the content starts with a shebang.
+    fn is_script(content: &str) -> bool {
+        content.starts_with("#!")
```

Comment:
be aware that shebang is not mandatory for shell scripts and also that after shebang can be something like python :)
In general as this means that shell scripts without shebang ends with "unknown file" it think it would be nice to combine both approaches...suffix and content. I would say ideally if there is suffix base decision on it as it is fast...and then check content if suffix is not there or unknown.

Reply:
Yes, it is optional, but you cannot trust the suffix (as the script comes from a URL, you might not even have a suffix). I think we can educate our users to include a shebang line :-)

Actually, I do not know if using something different from bash would work at all. Check [how we build the command](https://github.com/agama-project/agama/blob/master/rust/agama-cli/src/profile.rs).